### ISSUE TITLE
[C++20] [Coroutines] Use-After-Free problem if we destroy the coroutine handle in other thread while we're in await_suspend

### ISSUE THREAD

**Issue by ChuanqiXu9** [llvm/llvm-project#64152]
This comes from https://github.com/llvm/llvm-project/issues/56301. And this may be the later reported use-after-free problem.

The spec http://eel.is/c++draft/expr.await#5.1 says the coroutines are considered to be suspended once the `await_ready()` return false. In other words, the coroutines are considered to be suspended once it enters `await_suspend()`. Then it is legal to destroy a coroutine in other threads when it enters `await_suspend()` but not leaving yet.

But this is problematic and unimplementable actually, for example,

```C++
struct Awaiter {
  int suspended;

  bool await_ready() { return false; }

  void await_suspend(const std::coroutine_handle<> h) {
    init(suspended);
    
    func(h);

    use(suspended);
  }

  void await_resume() {}
  }
};
```

When we're executing the above `await_suspend()`, and we pass the coroutine handle to an async function `Register` then the coroutine handle may get destroyed. Then it is problematic to access the variable `suspended`.

Since the waiter lives on the coroutine frame but we may access it after the coroutine got destroyed in `await_suspend ()`.

I am not sure if this is specified in the spec. But I don't feel it is implementable in LLVM.

**Comment by jacobsa** [comment]
I think there is a typo in the example: what is `t`? Is that supposed to be `value`?

If so, I don't think anybody is claiming that this code should be legal. It's up to the implementer of the awaitable to ensure that it's not used after freeing.

The problem I [reported](https://github.com/llvm/llvm-project/issues/56301#issue-1289908373) as #56301 (and gave [another example of](https://github.com/llvm/llvm-project/issues/56301#issuecomment-1338605652)) is different. The C++ as written is legal; it's just that Clang turns a conditional write into an unconditional write.

**Comment by tzcnt** [comment]
I agree with jacobsa's interpretation. The developer should be aware that the awaiter may be destroyed after calling async_destroy_func(), and after that it is no longer safe to access variables of the awaiter or the coroutine promise.

Assuming that the developer's code as written is correct, clang needs to prevent certain optimizations that would move accesses from before async_destroy_func() to after it, as these may no longer be valid.
